From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tom Levy <tomlevy93@gmail.com>
Date: Tue, 19 May 2020 15:02:06 +1200
Subject: Remove repetitions of preferred options in affiliation dropdown

Since Symfony 4.4, the preferred choices of a ChoiceType are displayed
at the top and also in the main list (PR 32658).

It really doesn't suite the Affiliation dropdown, because the
preferred options ('No affiliation' and 'Add affiliation...') are
distinct from the main list of existing affiliations, and repeating
them at the start looks very silly.

Remove the repeated choices in the client, because there is no option
to control the ChoiceType's behaviour.
---
 webapp/templates/security/register.html.twig | 70 ++++++++++++++------
 1 file changed, 48 insertions(+), 22 deletions(-)

diff --git a/webapp/templates/security/register.html.twig b/webapp/templates/security/register.html.twig
index 9202ab327..743fc41f8 100644
--- a/webapp/templates/security/register.html.twig
+++ b/webapp/templates/security/register.html.twig
@@ -31,10 +31,57 @@
                 return;
             }
 
-            var $affiliationNew = $('[data-id="user_registration_affiliation_new"]');
+            // identify the placeholder, preferred options, separator, and repetitions
+
+            var $options = $affiliation.find('option');
+            var preferredOptions = [];
+            var preferredOptionValues = [];
+            var separator = null;
+            var repetitionsStart = null;
+            var repetitionsEnd = null;
+            for (var i = 0; i < $options.length; i++) {
+                var option = $options[i];
+                var $option = $(option);
+                if (separator === null) {
+                    // before separator
+                    if (i === 0 && $option.attr("value") === "") {
+                        // placeholder
+                        $option.addClass("placeholder");
+                    } else if (!$option.attr("value") && $option.is(':disabled')) {
+                        // separator
+                        separator = option;
+                        repetitionsStart = i + 1;
+                        repetitionsEnd = i + 1;
+                    } else {
+                        // preferred option
+                        preferredOptions.push(option);
+                        preferredOptionValues.push($option.attr("value"));
+                    }
+                } else {
+                    // after separator
+                    if (preferredOptionValues.indexOf($option.attr("value")) !== -1) {
+                        // repetition of a preferred option
+                        repetitionsEnd = i + 1;
+                    } else {
+                        // assume there are no more repetitions
+                        break;
+                    }
+                }
+            }
+            if (separator === null) {
+                preferredOptions = [];
+            } else {
+                $(preferredOptions).addClass("preferred");
+                // remove repetitions of the preferred options
+                if (repetitionsEnd > repetitionsStart) {
+                    $options.slice(repetitionsStart, repetitionsEnd).remove();
+                }
+            }
 
             // only display the new affiliation name and country fields when "Add affiliation..." is selected
 
+            var $affiliationNew = $('[data-id="user_registration_affiliation_new"]');
+
             if ($affiliationNew.length) {
                 $affiliation.on('change', function () {
                     var $affiliationNameGroup = $('#user_registration_affiliationName').closest('.form-group');
@@ -54,27 +101,6 @@
 
             // during search, move the options "No affiliation" and "Add affiliation..." to the end
 
-            var $options = $affiliation.find('option');
-            var preferredOptions = [];
-            var separator = null;
-            for (var i = 0; i < $options.length; i++) {
-                var option = $options[i];
-                var $option = $(option);
-                if (i === 0 && $option.attr("value") === "") {
-                    $option.addClass("placeholder");
-                } else if (!$option.attr("value") && $option.is(':disabled')) {
-                    separator = option;
-                    break;
-                } else {
-                    preferredOptions.push(option);
-                }
-            }
-            if (separator === null) {
-                preferredOptions = [];
-            } else {
-                $(preferredOptions).addClass("preferred");
-            }
-
             function matcher(params, data) {
                 if (params.term) {
                     var $element = $(data.element);
-- 
2.20.1

